### ISSUE TITLE
Use explicit instead of implicit conversion when it might throw

### ISSUE THREAD

**Issue by Fresa** [corvus-dotnet/Corvus.JsonSchema#231]
A lot of the primitives implement implicit conversion operators, `JsonString` to `string` for example, however they throw if the backing field prevents the conversion. This is generally [not recommended ](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/user-defined-conversion-operators#:~:text=Predefined%20C%23%20implicit%20conversions%20always%20succeed%20and%20never%20throw%20an%20exception.%20User%2Ddefined%20implicit%20conversions%20should%20behave%20in%20that%20way%20as%20well.%20If%20a%20custom%20conversion%20can%20throw%20an%20exception%20or%20lose%20information%2C%20define%20it%20as%20an%20explicit%20conversion.) and causes subtle conversion problems that are difficult to identify.

Example:
```
string? foo = null;
var bar = foo ?? JsonString.Undefined;
```
This throws InvalidOperationException.

I bet this is not something that is easy to fix at this point as this is such a central functionality. Maybe there is another simpler way to avoid it? An analyzer maybe?

**Comment by mwadams** [comment]
This is an interesting question and one we have had several discussions around.

It used to be the case that the implicit conversions never threw.

And then we decided that it was, after all, better to explode despite the recommendation.

We also had some considerable debate about whether to provide implicit conversions at all but it makes it all so much more generally usable that we left them in. @idg10 still thinks we have too many, I think!

I'm prepared to make this change at this stage because it doesn't require you to make *more checks*; it allows you to make fewer checks, and existing checks would still work.

**Comment by mwadams** [comment]
However my argument is that if the entity is not `IsValid => true` behaviour is undefined. So the exception is acceptable. If IsValid => True then you will never see this exception.

**Comment by Fresa** [comment]
`JsonString.Undefined` and `JsonString.Null` is weird, because none of them are json strings :) Maybe those should be defined as separate types (even though undefined is not a type) and use duck typing (maybe by using interfaces as you already do?)?

**Comment by idg10** [comment]
Our conceptual model for this family of types is that they are similar to `dynamic`. These inherently represent information which might have a shape at runtime that is not compatible with the shape your code expects. This directly reflects the reality of working with JSON data: unless you've taken steps to verify that it definitely conforms to the shape you expect, you don't know what it is.

Here's a `dynamic` example:

```cs
dynamic o = new ExpandoObject();
o.Foo = "hello";
int fi = o.Foo;
```

This compiles. You are allowed to assign `o.Foo` into a variable of `int`. This is similar in concept to assigning a `JsonString` into an `int`: this implies a runtime check that the underlying data is of the type required (which, in this case, it's not). In fact we can create an example that is even more similar to yours:

```cs
dynamic s = null!;
int v = s;
```

This fails at runtime with a complaint that `null` can't be converted to `int`.

Types such as `JsonString` are not quite as open as `dynamic`, in that they enable us to express expectations ahead of attempting to retrieve and change values. You could imagine a `dynamic<int>` indicating something we expect to be an `int`, but where that expectation might not be met at runtime. (That same expectation is embodied in the earlier example's final line, `int fi = o.Foo`, it's just that we're adding the idea that we could embody that expectation in a type.)

Since the basic concept here is that we are expressing expectations that might be dashed when we apply the code to actual data (because the inputs turned out not to have the shape we wanted) it follows that you can end up with self-exploding values - values that the compiler allows you to convert to some type such as `int`, but where that conversion might fail because a mismatch is detected at runtime.

We realise that this is...not a common way to represent this. But it's not unprecedented, in that `dynamic` has similar behaviour. And however you represent it, the basic tension with JSON and schemas is that you have reality and expectation, and unless you've performed a validation step to ensure that the reality aligns with the expectation, there might be a mismatch.

So we have a choice. We could say that getting values out of the world of `JsonString`, `JsonInteger`, and so on, always involved a method call, so your example would become:

```cs
string? foo = null;
var bar = foo ?? JsonString.Undefined.GetString();
```

In this formulation, the leap from "world of `JsonString/Integer/etc`" to "world of native .NET types" is explicit, so there's an obvious place for an exception to emerge. That's absolutely a possible way of doing this, and I think it would address your concern that implicit conversions shouldn't throw like this.

And you can code like this if you want, e.g.:

```cs
ja.As<JsonDateTime>().GetDateTime();
```

And there's also a `TryGetDateTime`. (We've realised there are some gaps though. There's a `TryGetString`, but no corresponding throwing `GetString` for example.)

But the problem with making this the _only_ way (and not supporting implicit conversions) is that if you have performed a validation step, it now becomes cumbersome to extract the information you want. So we have currently opted for the alternative in which conversions are implicit. If you call `IsValid` before trying to read the data you'll never see these exceptions. (OK, technically there are some JSON schemaisms where you can in fact contrive situations where a conversion might fail because there are some oddities around what numbers are in that world, but in general it won't happen.)

In the light of your feedback, we are considering some changes to exactly where conversions are available, and which ones are implicit. This might reduce the number of scenarios in which you will run into this, but it wouldn't change the fundamental tensions involved in dealing with statically typed representation of dynamically variable data.

**Comment by mwadams** [comment]
Another interesting thing raised by your example _is_ working as intended though! :)

```csharp
 var bar = foo ?? JsonString.Undefined;
```

`JsonString.Undefined` and `JsonString.Null` are types which represents "something that was expected to be a json string but was not present" and "something was expected to be a json string but was `null`.

These exist to help us map the tri-state of JSON ("present", "an instance of null", and "not present") into the C# world of "not null" and "null".

As part of this `JsonString.Undefined.IsValid()` and `JsonString.Null.IsValid()` are *always* false. Effectively, we define "known invalid" instances that can be used as singletons when you want to represent a null or an undefined instance of a type - and in fact, `default(T)` is equivalent to `T.Undefined`. 

There are extension methods for `IsNull()`, `IsNotNull()`, `IsUndefined()`, `IsNotUndefined()`, `IsNullOrUndefined()` and `IsNotNullOrUndefined()` to work with nullness and undefinedness in the json space.

But when you want to adapt back into the more native C# world, we provide `AsOptional<T>()` (and a special case `AsOptionalString()`)

 So, maybe what that code was trying to express was more like

```csharp
string? foo = null;
JsonString somePotentiallyNullOrUndefinedThing = JsonString.Undefined;

string bar = foo ?? somePotentiallyNullOrUndefinedThing .AsOptionalString(); // Won't compile
string? baz = foo ?? somePotentiallyNullOrUndefinedThing .AsOptionalString(); // Will compile
```

The first case will *not* compile, because `AsOptional()` produces a `string?` and that cannot be assigned to a `string` (saving you at compile time because you put in the verification step)

The second case will but is not necessarily useful!

So, having extracted a nullable string, you still end up writing code that looks like this:

```csharp
if (baz is string b)
{
   // work on my b
}
```

...and you've realized a string in the process.

Unless this is literally the last thing you do before handing that `string?` off into a dotnet library method that doesn't know about our Json extended types, it is probably better to stay in the JSON world and write code like this...

```csharp
JsonString somePotentiallyNullOrUndefinedThing = JsonString.Undefined;
if (somePotentiallyNullOrUndefinedThing.IsValid())
{
    // Carry on working with the json string
}
```

This avoids allocating the the dotnet string until it absolutely has to, and is guaranteed not to throw on conversion.

**Comment by mwadams** [comment]
As @idg10 says above - this is great real-world feedback, and we are definitely going to make some (breaking) changes as a result.

Ian - is there any reason we should move the design discussion to another thread and reference this one from it; or should we continue the design work in here? I think I prefer the former.

**Comment by mwadams** [comment]
We've started work on a significant simplification of the generated API.

You can take a look at the WIP at #246

**Comment by mwadams** [comment]
The first release of this is now available as v2.0.0.
